File: TopCoder/src/main/java/com/leetcode/PascalsTriangle.java
package com.leetcode;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

//https://leetcode.com/problems/pascals-triangle/
public class PascalsTriangle {

    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> list = new LinkedList<>();
        for (int i = 0 ; i < numRows ; i++) {
            if (list.size() == 0) {
                list.add(new LinkedList<>(Arrays.asList(1)));
            } else {
                List<Integer> lastList = list.get(list.size()-1);
                int newLength = lastList.size() + 1;
                List<Integer> newList = new LinkedList<>();

                for (int j = 0 ; j < newLength ; j++) {
                    int number1 = 0;
                    int number2 = 0;
                    if (j == lastList.size()) {
                        number2 = 0;
                    } else {
                        number2 = lastList.get(j);
                    }

                    if (j - 1 < 0) {
                        number1 = 0;
                    } else {
                        number1 = lastList.get(j-1);
                    }
                    newList.add(number1 + number2);
                }
                list.add(newList);
            }
        }
        return list;
    }

    public static void main(String[] args) {
        int n = 5;
        List<List<Integer>> result = new PascalsTriangle().generate(n);
        System.out.println(result);
    }
}

File: TopCoder/python/misc/PascalsTriangle.py
class Solution:
    def generate(self, numRows):
        #ret_val = []
        """
        for i in range(numRows):
            if i == 0:
                ret_val.append([1])
            else:
                last_list = ret_val[len(ret_val)-1]
                new_list = []
                for j in range(len(last_list)+1):
                    num1 = 0 if j == 0 else last_list[j-1]
                    num2 = 0 if j == len(last_list) else last_list[j]
                    new_list.append(num1+num2)
                ret_val.append(new_list)
        return ret_val
        """
        ret_val = [[1]*i for i in range(1, numRows+1)]

        for i in range(2, numRows):
            for j in range(1, i):
                ret_val[i][j] = ret_val[i-1][j] + ret_val[i][j-1]
        return ret_val

if __name__ == '__main__':
    sol = Solution()
    result = sol.generate(4)
    print(result)